src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2222,12 +2222,32 @@ public override BoundNode VisitWhileStatement(BoundWhileStatement node)
             return base.VisitWhileStatement(node);
         }
 
-        public override BoundNode VisitWithExpression(BoundWithExpression expr)
+        public override BoundNode VisitWithExpression(BoundWithExpression withExpr)
         {
-            // PROTOTYPE: This is wrong
-            SetResultType(expr,
-                expr.CloneMethod?.ReturnTypeWithAnnotations.ToTypeWithState()
-                    ?? TypeWithState.Create(expr.Type, NullableFlowState.NotNull));
+            Debug.Assert(!IsConditionalState);
+
+            var receiver = withExpr.Receiver;
+            VisitRvalue(receiver);
+            _ = CheckPossibleNullReceiver(receiver, checkNullableValueType: false);
+
+            var resultType = withExpr.CloneMethod?.ReturnTypeWithAnnotations ?? ResultType.ToTypeWithAnnotations();
+            var resultSlot = GetOrCreatePlaceholderSlot(withExpr);
+            // carry over the null state of members of 'receiver' to the result of the with-expression.
+            TrackNullableStateForAssignment(receiver, resultType, resultSlot, resultType.ToTypeWithState(), MakeSlot(receiver));
+            foreach (var expr in withExpr.InitializerExpression.Initializers)
+            {
+                if (expr is BoundAssignmentOperator assignment)
+                {
+                    VisitObjectElementInitializer(resultSlot, assignment);
+                }
+                else
+                {
+                    VisitRvalue(expr);
+                }
+            }
+
+            var placeholder = withExpr.InitializerExpression.Placeholder;
+            SetNotNullResult(placeholder);
```

Comment:
This is actually just copying what I saw in the object initializer visitors. I don't think I quite get what the Placeholder property is for. Also, it feels like the result type should be the return type of `Clone()`. This analysis probably doesn't account for a Clone with a nullable return type (need to add a test).